File: sdk/go/production_replay.go
package capsdk

import (
	"bytes"
	"errors"
	"sync"
	"time"
)

type ReplayOutcome uint8

const (
	ReplayOutcomeFirst ReplayOutcome = iota + 1
	ReplayOutcomeDuplicate
)

var (
	ErrReplayConflict         = errors.New("capsdk: replay digest conflict")
	ErrReplayStoreUnavailable = errors.New("capsdk: replay store unavailable")
)

type replayEntry struct {
	digest  []byte
	expires time.Time
}

type InMemoryReplayStore struct {
	mu      sync.Mutex
	entries map[string]replayEntry
}

func NewInMemoryReplayStore() *InMemoryReplayStore {
	return &InMemoryReplayStore{entries: make(map[string]replayEntry)}
}

func (s *InMemoryReplayStore) Admit(tenant, audience, sender string, messageID, digest []byte, expires time.Time) (ReplayOutcome, error) {
	if s == nil || tenant == "" || audience == "" || sender == "" || len(messageID) != 16 || len(digest) == 0 || !expires.After(time.Now()) {
		return 0, ErrReplayStoreUnavailable
	}
	key := tenant + "\x00" + audience + "\x00" + sender + "\x00" + string(messageID)
	s.mu.Lock()
	defer s.mu.Unlock()
	if existing, ok := s.entries[key]; ok && existing.expires.After(time.Now()) {
		if bytes.Equal(existing.digest, digest) {
			return ReplayOutcomeDuplicate, nil
		}
		return 0, ErrReplayConflict
	}
	s.entries[key] = replayEntry{digest: append([]byte(nil), digest...), expires: expires}
	return ReplayOutcomeFirst, nil
}

File: sdk/node/src/production-replay.ts
/** CAP-PRODUCTION atomic replay admission (in-memory reference store). */

export enum ReplayOutcome {
  First = "first",
  Duplicate = "duplicate",
}

export class ReplayConflictError extends Error {}
/** Callers MUST fail closed on this error, never treat it as first-seen. */
export class ReplayStoreUnavailableError extends Error {}

export interface ReplayStore {
  admit(
    tenant: string,
    audience: string,
    sender: string,
    messageId: Uint8Array,
    digest: Uint8Array,
    expiresAtMs: number,
  ): ReplayOutcome;
}

interface Entry {
  digest: Uint8Array;
  expiresAtMs: number;
}

function bytesEqual(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  for (let i = 0; i < a.length; i += 1) {
    if (a[i] !== b[i]) return false;
  }
  return true;
}

/** Reference ReplayStore. Durable deployments (e.g. Cordum) supply a
 * Redis- or other durable-backed implementation of the same interface. */
export class InMemoryReplayStore implements ReplayStore {
  private entries = new Map<string, Entry>();

  admit(
    tenant: string,
    audience: string,
    sender: string,
    messageId: Uint8Array,
    digest: Uint8Array,
    expiresAtMs: number,
  ): ReplayOutcome {
    if (!tenant || !audience || !sender || messageId.length !== 16 || digest.length === 0) {
      throw new ReplayStoreUnavailableError("invalid replay admission input");
    }
    const now = Date.now();
    if (expiresAtMs <= now) {
      throw new ReplayStoreUnavailableError("expiresAtMs must be in the future");
    }
    const key = `${tenant}\0${audience}\0${sender}\0${Buffer.from(messageId).toString("hex")}`;
    const existing = this.entries.get(key);
    if (existing && existing.expiresAtMs > now) {
      if (bytesEqual(existing.digest, digest)) {
        return ReplayOutcome.Duplicate;
      }
      throw new ReplayConflictError(`replay digest conflict for message_id=${Buffer.from(messageId).toString("hex")}`);
    }
    this.entries.set(key, { digest, expiresAtMs });
    return ReplayOutcome.First;
  }
}
